src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs

```diff
@@ -23,14 +23,57 @@ namespace Microsoft.AspNetCore.Routing;
 [DebuggerDisplay("Count = {Count}")]
 public class RouteValueDictionary : IDictionary<string, object?>, IReadOnlyDictionary<string, object?>
 {
-    // 4 is a good default capacity here because that leaves enough space for area/controller/action/id
-    private const int DefaultCapacity = 4;
     private static readonly ConcurrentDictionary<Type, PropertyHelper[]> _propertyCache = new ConcurrentDictionary<Type, PropertyHelper[]>();
 
-    internal KeyValuePair<string, object?>[] _arrayStorage;
+    internal DynamicArray<KeyValuePair<string, object?>> _arrayStorage;
     internal PropertyStorage? _propertyStorage;
     private int _count;
 
+    /// <summary>
+    /// Creates a new <see cref="RouteValueDictionary"/> from the provided Span of items.
+    /// </summary>
+    /// <param name="items">The items spans.</param>
+    /// <returns>A new <see cref="RouteValueDictionary"/>.</returns>
+    public static RouteValueDictionary FromValues(Span<KeyValuePair<string, object?>> items)
+    {
+        // We need to compress the array by removing non-contiguous items. We
+        // typically have a very small number of items to process. We don't need
+        // to preserve order.
+        var start = 0;
+        var end = items.Length - 1;
+
+        // We walk forwards from the beginning of the array and fill in 'null' slots.
+        // We walk backwards from the end of the array end move items in non-null' slots
+        // into whatever start is pointing to. O(n)
+        while (start <= end)
+        {
+            if (items[start].Key != null)
+            {
+                start++;
+            }
+            else if (items[end].Key != null)
+            {
+                // Swap this item into start and advance
+                items[start] = items[end];
+                items[end] = default;
+                start++;
+                end--;
+            }
+            else
+            {
+                // Both null, we need to hold on 'start' since we
+                // still need to fill it with something.
+                end--;
+            }
+        }
+
+        return new RouteValueDictionary
+        {
+            _arrayStorage = new(items[..start]),
+            _count = start,
+        };
```

Comment:
Could this instead be:
```C#
int next = 0;
for (int i = 0; i < items.Length; i++)
{
    if (items[i].Key is not null)
    {
        if (next != i) items[next] = items[i];
        next++;
    }
}

return new RouteValueDictionar
{
    _arrayStorage = new(items[..next]),
    _count = next,
}
```
?

Reply:
I didn't write the original code, just copied it, but I'll see if this works.